File: src/data/genPhotos.R
path <- getwd()
if (substr(path, nchar(path), nchar(path)) != "/") {
  path <- paste0(path, "/")
}
includePath <- "include/"

source(paste0(path, includePath, 'generateKeggPathwayImages.R'))

user_rdata_file <- "../../data/rdata/countsAndSamples.RData"
url <- "https://ftp.ebi.ac.uk/pub/databases/gencode/Gencode_human/release_19/gencode.v19.chr_patch_hapl_scaff.annotation.gtf.gz"
dest <- "../../data/raw/gencode.v19.chr_patch_hapl_scaff.annotation.gtf.gz"

dir.create("../../data/raw", recursive = TRUE, showWarnings = FALSE)

user_gtf_file <- "../../data/raw/gencode.v19.chr_patch_hapl_scaff.annotation.gtf"
if (!file.exists(user_gtf_file)) {
  if (!file.exists(dest)) {
    download.file(url, dest, mode = "wb")
  }
  R.utils::gunzip(dest, remove = TRUE)
} else {
  message("GTF file already exists, skipping download and extraction.")
}
source(paste0(path, includePath, "statisticalAnalysis.R"))
source(paste0(path, includePath, "annotateData.R"))

message(paste("Loading data from:", user_rdata_file))
if (!file.exists(user_rdata_file)) {
  stop("The .RData file was not found. Check the path in 'user_rdata_file'.")
}
load(user_rdata_file)
if (!exists("countsRaw")) {
  stop("The loaded .RData file does not contain an object named 'countsRaw'.")
}
message("Normalizing data...")
dataFiltered <- normalizeDESeq2NoReport(countsRaw)
message("Normalization completed.")

annotation_result <- annotateData(user_gtf_file, dataFiltered)
dataFiltered_annotated <- annotation_result$dataFiltered
message("Annotation completed.")


sampleGroups <- NULL
if (exists("sampleInfo") && "Group" %in% colnames(sampleInfo)) {
  if (nrow(sampleInfo) == ncol(dataFiltered_annotated)) {
    sampleGroups <- sampleInfo$Group
  } else {
    warning("sampleInfo$Group omitted: number of rows does not match number of samples")
  }
} else {
  message("No sampleInfo$Group – files will be named by sample only")
}


message("Starting generation of KEGG pathway images...")
matrixPath <- generateKeggPathwayImages("../../data/", dataFiltered_annotated, sampleGroups = sampleGroups)

message(paste("Completed. Matrices saved to:", matrixPath))
message(paste("Images saved to:", gsub("Matrices", "Images", matrixPath)))

File: src/data/include/genPhotos.py
import os
import gzip
import shutil
import requests
import pandas as pd

from normalize_deseq import normalize_deseq2_no_report
from annotateData import annotateData
from generatePathwayImages import generate_kegg_pathway_images


# ==============================
# Paths
# ==============================

BASE_DIR = os.path.abspath(os.path.dirname(__file__))

data_dir = os.path.abspath(os.path.join(BASE_DIR, "../../data"))
raw_dir = os.path.join(data_dir, "raw")
output_dir = os.path.join(data_dir, "output")
script_dir = os.path.dirname(os.path.abspath(__file__))


os.makedirs(raw_dir, exist_ok=True)
os.makedirs(output_dir, exist_ok=True)

# ==============================
# Files
# ==============================

counts_file_csv = "counts_raw.csv"
#counts_file_tsv = os.path.join(raw_dir, "counts_raw.tsv")
#counts_file = counts_file_csv if os.path.exists(counts_file_csv) else counts_file_tsv
counts_file = counts_file_csv

gtf_url = "https://ftp.ebi.ac.uk/pub/databases/gencode/Gencode_human/release_19/gencode.v19.chr_patch_hapl_scaff.annotation.gtf.gz"
gtf_gz = os.path.join(raw_dir, "gencode.v19.annotation.gtf.gz")
gtf_file = os.path.join(raw_dir, "gencode.v19.annotation.gtf")

kegg_json = os.path.join(script_dir, "kegg_pathways.json")

# ==============================
# Download GTF if needed
# ==============================

if not os.path.exists(gtf_file):
    print("Downloading GTF...")
    if not os.path.exists(gtf_gz):
        response = requests.get(gtf_url, stream=True)
        with open(gtf_gz, "wb") as f:
            shutil.copyfileobj(response.raw, f)

    print("Extracting GTF...")
    with gzip.open(gtf_gz, "rb") as f_in:
        with open(gtf_file, "wb") as f_out:
            shutil.copyfileobj(f_in, f_out)

    os.remove(gtf_gz)
else:
    print("GTF file already exists, skipping download.")

# ==============================
# Load raw counts
# ==============================

if not os.path.exists(counts_file):
    raise FileNotFoundError("Raw counts file not found.")

print(f"Loading raw counts from: {counts_file}")

if counts_file.endswith(".csv"):
    counts_raw = pd.read_csv(counts_file, index_col=0)
else:
    counts_raw = pd.read_csv(counts_file, sep="\t", index_col=0)

counts_raw = counts_raw.apply(pd.to_numeric, errors="coerce").dropna()

# ==============================
# Normalize
# ==============================

print("Normalizing data...")
data_filtered = normalize_deseq2_no_report(counts_raw, fast=False)
print("Normalization completed.")

# ==============================
# Annotate
# ==============================

print("Annotating data...")
annotation_result = annotateData(gtf_file, data_filtered)
data_filtered_annotated = annotation_result["dataFiltered"]
print("Annotation completed.")

# ==============================
# Sample groups (optional)
# ==============================

sample_groups = None
sample_info_path = os.path.join(raw_dir, "sample_info.csv")

if os.path.exists(sample_info_path):
    sample_info = pd.read_csv(sample_info_path)
    if "Group" in sample_info.columns:
        if len(sample_info) == data_filtered_annotated.shape[1]:
            sample_groups = sample_info["Group"].tolist()
        else:
            print("Warning: sample_info length mismatch – ignoring groups.")
else:
    print("No sample_info.csv found – files will be named by sample only.")

# ==============================
# Generate KEGG images
# ==============================

print("Starting generation of KEGG pathway images...")

matrix_path = generate_kegg_pathway_images(
    path=data_dir,
    data_filtered=data_filtered_annotated,
    json_path=kegg_json,
    sample_groups=sample_groups
)

print(f"Completed.")
print(f"Matrices saved to: {matrix_path}")
print(f"Images saved to: {matrix_path.replace('Matrices', 'Images')}")